structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,341 @@
+		if match := before.FindStringSubmatch(envVar.Value); match != nil {
+			// first entry is the leftmost substring: second entry is the first group
+			currentVar = match[1]
+		}
+		if match := after.FindStringSubmatch(envVar.Value); match != nil {
+			currentVar = match[1]
+		}
+		if currentVar != "" {
+			if val, ok := imageEnv[currentVar]; ok {
+				env = append(env, envVar.Key+"="+strings.Replace(envVar.Value, "$"+currentVar, val, -1))
+			} else {
+				t.Errorf("Variable %s not found in image env! Check test config.", currentVar)
+			}
+		} else {
+			env = append(env, envVar.Key+"="+envVar.Value)
+		}
+	}
+	// since these are global envvars, just overwrite the original image
+	d.originalImage = d.runAndCommit(t, env, nil)
+	d.currentImage = d.originalImage
+	return nil
+}
+
+func (d *DockerDriver) ResetEnvVars(t *testing.T, vars []unversioned.EnvVar) {
+	// since the container will be destroyed after the tests, this is a noop
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d *DockerDriver) retrieveFile(t *testing.T, path string, directory bool) (string, error) {
+	ctx := context.Background()
+
+	// TODO(nkubala): this contains a hack to get around the fact that
+	// the client will not allow creating a container without a command.
+	// we should remove this, or at the very least find a better alternative
+	// given that not every container is guaranteed to have a shell.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Cmd:          []string{"NOOP_COMMAND_DO_NOT_RUN"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	if directory {
+		return "", nil
+	} else {
+		tmpFile, err := ioutil.TempFile("", "structure_test")
+		if err != nil {
+			t.Errorf("Error when creating temp file: %s", err.Error())
+			return "", err
+		}
+		stream := bufio.NewWriter(tmpFile)
+
+		err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+			OutputStream: stream,
+			Path:         path,
+			Context:      ctx,
+		})
+		if err != nil {
+			t.Errorf("Error when downloading file from container: %s", err.Error())
+			return "", err
+		}
+		stream.Flush()
+		tmpFile.Close()
+		return tmpFile.Name(), nil
+	}
+}
+
+func (d *DockerDriver) StatFile(t *testing.T, path string) (os.FileInfo, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+
+	f, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		return nil, nil
+	}
+	return f, nil
+}
+
+func (d *DockerDriver) ReadFile(t *testing.T, path string) ([]byte, error) {
+	file, err := d.retrieveFile(t, path, false)
```

Comment:
Can retrieveFile avoid writing to disk?

Reply:
hmm, since the result of this is used in a call to `os.Stat()`, I'm not sure it's possible. we use the result of `os.Stat()` to look at metadata about the file such as permission, etc. thoughts?